test/integration/framework/etcd.go

```diff
@@ -168,6 +169,9 @@ func RunCustomEtcd(dataDir string, customFlags []string) (url string, stopFn fun
 
 // EtcdMain starts an etcd instance before running tests.
 func EtcdMain(tests func() int) {
+	// Bail out early when -help was given as parameter.
+	flag.Parse()
+
```

Comment:
EtcdMain is called in TestMain functions, which i would assume is called even if -help is passed?

i wonder if there are more more elegant ways to do this, since the framework should not be importing flag maybe.
would it be a problem if flag.Parse is called in multiple places?

perhaps instead of calling Parse we can lookup flags that prevent test execution and return here.

Reply:
> EtcdMain is called in TestMain functions, which i would assume is called even if -help is passed?

Yes, that's how we get here. All of the `TestMain` functions are wrappers around `framework.EtcdMain`. `flag.Parse` could be moved into the various `TestMain`, but that's then 54 files that need to be changed instead of one.

As `framework.EtcdMain` is the `TestMain` for integration tests, I think it is okay to also do the common flag handling here.

> perhaps instead of calling Parse we can lookup flags that prevent test execution and return here.
(but still imports flags obviously...)

That smells like a hack.

Reply:
looks like there is already one place that calls flag.Parse before EtcdMain.
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/test/integration/scheduler_perf/main_test.go#L27-L28

also, the top level flag.Parse() is calling Parse on the  FlagSet  - CommandLine.
looking at the go source code the function modifies state in the FlagSet:
https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/flag/flag.go;l=1005;drc=refs%2Ftags%2Fgo1.17.1
https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/flag/flag.go;drc=refs%2Ftags%2Fgo1.17.1;l=923
which might be OK, but still calling it multiple times doesn't seem right.

> Yes, that's how we get here. All of the TestMain functions in integration tests are wrappers around framework.EtcdMain. flag.Parse could be moved into the various TestMain, but that's then 54 files that need to be changed instead of one.

i think the right thing to do is to call flag.Parse once per TestMain, before reaching EtcdMain.

Reply:
> looks like there is already one place that calls flag.Parse before EtcdMain.

This could be removed.

> i think the right thing to do is to call flag.Parse once per TestMain, before reaching EtcdMain.

Really replicate the same code at *every* caller of EtcdMain, instead of once *in* EtcdMain? Sorry, to me that still sounds like useless code duplication.

> ...or just call it only once, in a centralized location - e.g. the entry point of the binary that accepts -h.

TestMain is the entry point of a test binary. There is no other.

Reply:
> Really replicate the same code at every caller of EtcdMain, instead of once in EtcdMain? Sorry, to me that still sounds like useless code duplication.

it is indeed a lot of code duplication, it just doesn't seem write to do flag parsing in the function that  "starts an etcd instance before running tests" (EtcdMain). i guess, we can proceed with the idea but clean existing locations that already have Parse in TestMain.

@oomichi do you have an opinion on this flag.Parse in EtcdMain topic?

Reply:
> i guess we can proceed with the idea but clean existing locations that already have Parse in TestMain.

Done. There was only the one from https://github.com/kubernetes/kubernetes/pull/89139